base/applications/mspaint/mouse.cpp

```diff
@@ -58,216 +61,444 @@ BOOL nearlyEqualPoints(INT x0, INT y0, INT x1, INT y1)
+        else
+            SetPixel(m_hdc, x, y, m_bg);
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnButtonUp(bLeftButton, x, y);
+    }
+
+    void OnCancelDraw()
+    {
+        OnButtonUp(FALSE, 0, 0);
+        imageModel.Undo();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_BRUSH
+struct BrushTool : ToolBase
+{
+    BrushTool() : ToolBase(TOOL_BRUSH)
+    {
+    }
+
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            Brush(m_hdc, last.x, last.y, x, y, m_fg, toolsModel.GetBrushStyle());
+        else
+            Brush(m_hdc, last.x, last.y, x, y, m_bg, toolsModel.GetBrushStyle());
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+        draw(bLeftButton, x, y);
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnButtonUp(bLeftButton, x, y);
+    }
+
+    void OnCancelDraw()
+    {
+        OnButtonUp(FALSE, 0, 0);
+        imageModel.Undo();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_AIRBRUSH
+struct AirBrushTool : ToolBase
+{
+    AirBrushTool() : ToolBase(TOOL_AIRBRUSH)
+    {
+    }
+
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            Airbrush(m_hdc, x, y, m_fg, toolsModel.GetAirBrushWidth());
+        else
+            Airbrush(m_hdc, x, y, m_bg, toolsModel.GetAirBrushWidth());
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+        draw(bLeftButton, x, y);
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        draw(bLeftButton, x, y);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnCancelDraw()
+    {
+        OnButtonUp(FALSE, 0, 0);
+        imageModel.Undo();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_TEXT
+struct TextTool : ToolBase
+{
+    TextTool() : ToolBase(TOOL_TEXT)
+    {
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        POINT temp;
+        if (bLeftButton)
```

Comment:
Note: at least with Windows 10 paint, we can indifferently use mouse-left or mouse-right to create a text zone.

Reply:
OK.